presto-common/src/main/java/com/facebook/presto/common/block/ArrayBlockBuilder.java

```diff
@@ -337,4 +338,39 @@ public String toString()
     {
         return format("ArrayBlockBuilder(%d){positionCount=%d}", hashCode(), getPositionCount());
     }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        if (currentEntryOpened) {
+            return false;
+        }
+        ArrayBlockBuilder other = (ArrayBlockBuilder) obj;
+        return this.positionCount == other.positionCount &&
+                Arrays.equals(this.valueIsNull, other.valueIsNull) &&
+                Objects.equals(this.values, other.values) &&
+                Arrays.equals(this.offsets, other.offsets) &&
```

Comment:
valueIsNull, values and offsets array are overallocated. Two different map block builders can grow at different capacity, depending on the intiailEntryCount. 

so valuesIsNull and values shoudl be compared until the length positionCount.
offsets are allocated one more, so they should be compared until positionCount + 1.

retainedSizeInBytes should not be included in the comparison as they can have different sizes.

Reply:
I am including as many elements here for correctness, as we may end up with silent failures with false positive. Also the implementation here conforms with the corresponding ArrayBlock.